modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/splitter/Fragment.java

```diff
@@ -75,39 +87,64 @@ public void init(PlannerContext ctx, RelMetadataQuery mq) {
                 IgniteReceiver receiver = input.left;
                 RelSource source = input.right;
 
-                source.init(mapping, receiver.distribution(), ctx, mq);
+                source.bindToTarget(new RelTargetImpl(id, fragmentMapping(ctx, info, mq), receiver.distribution()), ctx, mq);
             }
         }
     }
 
-    /** {@inheritDoc} */
-    @Override public long exchangeId() {
-        return exchangeId;
-    }
-
-    /** {@inheritDoc} */
-    @Override public void init(NodesMapping mapping, IgniteDistribution distribution, PlannerContext ctx, RelMetadataQuery mq) {
-        assert remote();
-
-        ((IgniteSender) root).target(new RelTargetImpl(exchangeId, mapping, distribution));
-
-        init(ctx, mq);
-    }
-
     /**
      * @return Root node.
      */
     public RelNode root() {
         return root;
     }
 
+    /** {@inheritDoc} */
+    @Override public long fragmentId() {
+        return id;
+    }
+
     /** {@inheritDoc} */
     @Override public NodesMapping mapping() {
         return mapping;
     }
 
+    /** {@inheritDoc} */
+    @Override public void bindToTarget(RelTarget target, IgniteCalciteContext ctx, RelMetadataQuery mq) {
+        assert !local();
+
+        ((IgniteSender) root).target(target);
+
+        init(ctx, mq);
+    }
+
     /** */
-    private boolean remote() {
-        return root instanceof IgniteSender;
+    public boolean local() {
+        return !(root instanceof IgniteSender);
```

Comment:
Not obvious naming. Why sender in the root related to local fragment? Please add a comment o rename method to something more meaningful (which is a better solution).

Reply:
NOT sender node in the root of fragment indicates what itis a local part of distributed query.